### DISCUSSION TITLE
Need help with differences between similar legacy lpadmin options

### DISCUSSION THREAD

**Discussion by bfg01** [OpenPrinting/cups#707] [Q&A] [answered]
Due to my ignorance in regards of where to ask, I wrongly asked this as an "issue" in CUPS.
Now I'd like to retry here in the proper place.

At office I deal with both fairly recent and ancient printers.

The later ones can only work with old PPDs; this is, by being added with lpadmin -i or -P option.
The thing is, even after reading the manual and docs, I simply don't understand what the actual difference is and when each one used to be used, even if in strictly "historic context".

Aside of this, for cancelling failed print jobs, there are (or were?) the commands "lprm" and "cancel".

In both cases, what is or was the difference? Which one to use when?

And by quoting from the answer in the issue:
"both options will be removed in CUPS 3.0 and you will have to use a Printer Application for your classic (ancient) printers"
What is a "Printer Application" in this context?

Thanks beforehand for your attention.

**Comment by zdohnal** [comment, marked as answer]
Hi, 

> The later ones can only work with old PPDs; this is, by being added with lpadmin -i or -P option. The thing is, even after reading the manual and docs, I simply don't understand what the actual difference is and when each one used to be used, even if in strictly "historic context".

**-i** was for System V interface scripts (scripts which were run at the end of job) and its usage with PPD was a side-effect. **-P** is for providing PPD files only. This is described in ```man lpadmin``` - to sum it up, both are meant for a different file types.

> 
> Aside of this, for cancelling failed print jobs, there are (or were?) the commands "lprm" and "cancel".
> 
> In both cases, what is or was the difference? Which one to use when?

Each of them are based on a different Unix version - ```lprm``` is for Berkeley, ```cancel``` is for System V - AFAIK there are no differences in functionality (the parameters names may differ, but the functionality stays the same/similar) - so it is a matter of your preference what to use.

> 
> And by quoting from the answer in the issue: "both options will be removed in CUPS 3.0 and you will have to use a Printer Application for your classic (ancient) printers" 

Your quote is missing a important part from the original issue - Brian speaks there about **-i** and **-P**, which are going to be removed in the future.

>What is a "Printer Application" in this context?
> 

To put it simple, printer application is new generation of driver - another project/package separated from CUPS, which gives user support for old printers (retrofitting printer applications) or enhanced support for current printers, if common standards (IPP Everywhere, Airprint, Mopria...) do not support all capabilities of the printer (native printer applications). The printer application is capable to communicate with CUPS via standardized protocol and can report itself via mDNS, so CUPS picks it up automatically without any installation.

> Thanks beforehand for your attention.

**Comment by bfg01** [comment]
Thanks sir for your explanation.

I searched documentation about "Printer Application", and all coincided that, in the end, it's a sandbox (snap) application actually provided by the manufacturers themselves through their respective websites.

But what if manufacturer stopped supporting some models for good?

For example, for a Canon IR-ADV 4245 (NOT the 4245i !), I couldn't find any "snap" package last time I checked the website...

Thanks again.

**Reply by zdohnal** [reply]
If the old driver is open source and shipped in Linux distribution, we in OpenPrinting have retrofitting printer application to support it.
In case the manufacturer does  provide only the classic driver and no printer application for the printer, you can use Legacy printer application from OpenPrinting/pappl-retrofit - this printer application won't be containerized, so it depends on your distribution whether it ships it or not.

But either way, all new drivers should be implemented as a printer applications and the existence of legacy printer application is only for retrofitting support of old printers.

**Reply by debiantriage** [reply]
> For example, for a Canon IR-ADV 4245 (NOT the 4245i !), I couldn't find any "snap" package last time I checked the website...

Obtaining information from Canon about this device is akin to finding out how to extract hen's teeth! [Eventually](https://www.usa.canon.com/support/p/imagerunner-advance-4245):

> 

Functions |  Optional: Print (PCL, Adobe® PostScript® 3&trade;),...
-- | --

PostScript is an option. Canon has pushed UFR II as its standard offering on this printer. Fair enough, but it leaves a 4245 user with some work to do when CUPS 3.0 comes along.

The Legacy Printer Application should deal with an existing UFR II setup. Problems when testing? Submit a bug report to OpenPrinting.

**Comment by bfg01** [comment]
So for the aforementioned 4245, I found PostScript snapshot at OpenPrinting website.
Wanted to test, but didn't work: any document tried, it just prints a first sheet with few random text, then literally endless sheets with nothing printed, until I cancel the print job.

Called the technician and after inspecting the printer's settings, found that seemingly this particular model does not have {Ghost,Post}Script at all, only UFR II and PCL 5e/6, and that the one actually having support was the 'i' variant, the 4245i (which also has the IPP support).

So, 2 questions:

---What do I normally do with these printer cases?

---When working with other old printers with working Printer Application snap, how do I actually use it?
Docs say in general "install snap and add printer via web interface".
Do I forcibly need to install with snap command, or can I just run the downloaded snap?
Is web interface forcibly needed, or can I just use the lpadmin way and the URI, just like with IPP Everywhere?

**Reply by zdohnal** [reply]
> ---What do I normally do with these printer cases?

Use ghostscript-printer-app, because once you read the README, you see it supports PCL5e and PCL6, in case you are fine with generics.

Or, I hoped it was clear from the last reply because you mentioned looking for driver on Canon website and not founding a printer application there, so I assumed you took the driver from there, use Legacy Printer Application from pappl-retrofit with the old driver.

> 
> ---When working with other old printers with working Printer Application snap, how do I actually use it? Docs say in general "install snap and add printer via web interface". Do I forcibly need to install with snap command, or can I just run the downloaded snap?

I'm a basic Snap user and I only did it via installing the printer application via ```snap``` command, because the printer applications are available in Snapcraft repos, so the downloading+installing is done via snap command. From what I saw upstreams put their snaps/flatpaks to some common repository (Snapcraft in case of snaps) where users can install it via command.
In case you choose to use Legacy Printer Application you don't need to know Snaps, as I wrote in my previous comment - due its nature it will be shipped in classic distribution systems, so it depends whether your distro will package pappl-retrofit.

> Is web interface forcibly needed, or can I just use the lpadmin way and the URI, just like with IPP Everywhere?

Yes, you can do the same thing via CLI, but you use printer application's name as command f.e. ps-printer-app. See ps-printer-app --help, and then you use lpadmin (in case you don't want to use mDNS or need to share the printer further) to install  the printer permanently. In case you are fine with mDNS and you're not on server, you don't have to use lpadmin or CUPS Web Interface at all (considering your applications support 7 years old CUPS API).

**Comment by debiantriage** [comment]
Perhaps I can offer this [Debian wiki page](https://wiki.debian.org/CUPSNewArchitecture) as a gentle (and hopefully accurate) introduction to the topics raised here. @bfg01  What OS are you using?

It is important to appreciate that, as Zdenek has said, _almost_  all **free** classic PPDs and drivers are available in the ps-printer-app, hplip-printer-app, gutenprint-printer-app and ghostscript-printer-app Printer Applications and Debian will presumably deliver them to users in the usual .deb format.

For _non-free_ classic drivers there is the Legacy Printer Application. It cannot be snapped as it requires access to things outside the snap package. pappl-retrofit would require compilation to experience it. I am rubbish at compiling, so haven't tested it. Whether vendors such as Brother and Canon will contribute to the snap store remains to be seen. It's up to them; _we_ have all bases covered.

**Comment by bfg01** [comment]
@debiantriage More than complete answer aside that of the maintainer.

Thanks very much sirs. I think I finally learned about the whole topic about upcoming CUPS 3.
